lib/firenest/replicated_state/server.ex

```diff
@@ -0,0 +1,288 @@
+defmodule Firenest.ReplicatedState.Server do
+  @moduledoc false
+  use Firenest.SyncedServer
+
+  alias Firenest.SyncedServer
+  alias Firenest.ReplicatedState.{Store, Remote, Handler}
+
+  def child_spec({name, topology, handler, opts}) do
+    server_opts = [name: name, topology: topology]
+
+    %{
+      id: name,
+      start: {SyncedServer, :start_link, [__MODULE__, {name, handler, opts}, server_opts]}
+    }
+  end
+
+  @impl true
+  def init({name, handler, opts}) do
+    Process.flag(:trap_exit, true)
+
+    store = Store.new(name)
+
+    delayed_fun = &Process.send_after(self(), {:update, &1, &2, &3}, &4)
+    {handler, server_opts} = Handler.new(handler, opts, delayed_fun)
+
+    broadcast_timeout = Keyword.get(opts, :broadcast_timeout, 50)
+    broadcast_fun = fn -> Process.send_after(self(), :broadcast_timeout, broadcast_timeout) end
+
+    remote_changes = Keyword.get(server_opts, :remote_changes, :ignore)
+    max_deltas = Keyword.get(opts, :max_remote_deltas, 5)
+    remote = Remote.new(remote_changes, broadcast_fun, max_deltas)
+
+    {:ok,
+     %{
+       store: store,
+       handler: handler,
+       remote: remote
+     }}
+  end
+
+  @impl true
+  def handshake_data(%{remote: remote}), do: Remote.clock(remote)
+
+  @impl true
+  def handle_call({:put, key, pid, arg}, _from, state) do
+    %{store: store, handler: handler, remote: remote} = state
+
+    link(pid)
+
+    unless Store.present?(store, key, pid) do
+      case Handler.local_put(handler, arg, key, pid) do
+        {:put, value, delta, handler} ->
+          remote = Remote.local_put(remote, key, pid, value)
+          store = Store.local_put(store, key, pid, value, delta)
+          {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+        {:delete, value, _delta, handler} ->
+          remote = Remote.local_put(remote, key, pid, value)
+          remote = Remote.local_delete(remote, key, pid)
+          handler = Handler.local_delete(handler, [value])
+          {:reply, :ok, %{state | handler: handler, remote: remote}}
+      end
+    else
```

Comment:
unless/else is usually an anti-pattern. I am fine with it but I thought I would double check.

Reply:
We should kill unless+else with fire :)